build/__init__.py

```diff
@@ -106,6 +106,10 @@ def __init__(self, srcdir='.'):  # type: (str) -> None
         self.hook = pep517.wrappers.Pep517HookCaller(self.srcdir, self._backend,
                                                      backend_path=self._build_system.get('backend-path'))
 
+    @property
+    def build_dependencies(self):  # type: () -> Set[str]
+        return set(self._build_system.get('requires', []))
```

Comment:
Hm... I am a bit surprised by this. I suppose it's fair, but I was originally thinking that *any* missing fields in the `[build-system]` table would come from the defaults, in which case you should be able to use `self._build_system["requires"]`).

The way it is now, you supply a default value if no `[build-system]` table is provided, but if `[build-system]` is only partially specified, you leave it fully unspecified.

The thing is, there was a decently long era (6 months-1 year) where `pip` supported PEP 518 but not PEP 517, and there are some people who only care about PEP 518 and not PEP 517, so I wouldn't be surprised if there are a bunch of projects looking like this:

```toml
[build-system]
requires = ["Cython >= 28.0", "setuptools >= 48.0.3", "wheel"]
```

Or something of that nature, which would leave the build-backend unspecified! I recognize that it's a judgement call, but I was thinking that instead of [this code for setting the defaults](https://github.com/FFY00/python-build/blob/8db9c1110fb1d965d95326daf06a96a136e76d6b/build/__init__.py#L88), you'd want something more like this:

```python
self._build_system = {
    'build-backend': 'setuptools.build_meta:__legacy__',
    'requires': [
        'setuptools >= 40.8.0',
        'wheel'
    ]
}
self._build_system.update(self._spec.get('build-system', {}))
```

Though probably it's safe to figure out exactly what `pip` does and do that.

Reply:
Where does `Cython >= 28.0` come from? Is it tied to the `setuptools` requirement? If so, I don't believe we should default to that.

When `build-backend` is missing we default to:
```toml
[build-system]
requires = ['setuptools >= 40.8.0', 'wheel']
build-backend = 'setuptools.build_meta:__legacy__'
```

But I believe it's wrong to default to any `requires` **if** `build-backend` is specified. I could have an in-tree backend and not specify any `requires`, in this case I wouldn't expect a user to add `requires = []`.

We could select the default value for `requires` based on `build-backend`. For `build-backend = 'setuptools.build_meta:__legacy__'` we know that `requires = ['setuptools >= 40.8.0', 'wheel']`. But I do not believe this would be the right approach, it's a slippery slope, who decides which backends we will and will not provide default values for?

For the other way round, I am not sure. I would be okay with defaulting `build-backend` to `setuptools.build_meta:__legacy__` and appending `['setuptools >= 40.8.0', 'wheel']` to ``requires.

Reply:
> Where does `Cython >= 28.0` come from? Is it tied to the `setuptools` requirement? If so, I don't believe we should default to that.

Sorry, I should have been more clear. There are a bunch of packages out there that specify `build-system.requires` but not `build-system.build-backend`, because if you don't specify `build-system.build-backend`, `pip` will default to `setuptools`. Some people deliberately don't set it because they want `__legacy__`, and other people just never set it because they set up their `pyproject.toml` before `pip` had any support for the `build-backend` key.

> But I believe it's wrong to default to any `requires` **if** `build-backend` is specified. I could have an in-tree backend and not specify any `requires`, in this case I wouldn't expect a user to add `requires = []`.

This is a matter of taste, I suppose.  I think that you *are* supposed to explicitly specify `requires = []` and the behavior is undefined-ish if you don't. PEP 518 says that if the table exists but the `requires` field is missing the tool [should consider it an error](https://www.python.org/dev/peps/pep-0518/#build-system-table).

I think `requires = []` is kind of a bad default, because almost no one has in-tree backends, and there's actually [a separate mechanism for declaring that you have one](https://www.python.org/dev/peps/pep-0517/#in-tree-build-backends), so anyone who hasn't specified the `backend-path` field *must* have dependencies.

That said, it is definitely more important to have a default value for an unspecified `build-backend` if `requires` is specified than to have a default value for `requires` if `build-backend` is specified. I think if `build-backend` is specified without `requires`, the reasonable things to do are to default to the normal default value of `requires` or to throw an exception. The arguments in favor of choosing the "populate a default and use `.upgrade`"-style strategy are:

1. A lot of tools just assume that `setuptools` and `wheel` are present, even now. People who need them might fail to specify it, but people who need anything *else* will definitely specify it.
2. If the person actually wanted `requires=[]`, it probably doesn't actively break their build to also install `setuptools` and `wheel` (and if it does, they can add `requires=[]`), so it's a much softer kind of failure.
3. It's easier for end users to understand the defaulting behavior if it follows a simple rule: "Anything missing from the `build-system` table is taken from these defaults." is much easier to understand than, "If you specify `requires` then `build-backend` comes from this default, but if you specify `build-backend` then `requires` is something different, and if the table is missing entirely it's these default values".

Reply:
> This is a matter of taste, I suppose. I think that you _are_ supposed to explicitly specify `requires = []` and the behavior is undefined-ish if you don't. PEP 518 says that if the table exists but the `requires` field is missing the tool [should consider it an error](https://www.python.org/dev/peps/pep-0518/#build-system-table).

Then I think we should consider it an error :)
I would like to be as strict as possible to avoid leaking implementation details. The problem is people complaining "it works in `pip` but doesn't work with `build`". The project description says we are a "*correct* PEP517 package builder" so I think it makes sense to go down that route, what do you think?

> I think if `build-backend` is specified without `requires`, the reasonable things to do are to default to the normal default value of `requires` or to throw an exception.

I agree. I vote to throw an exception.

Reply:
This should be resolved, @pganssle can you check if the implementation is okay?